test/iteration.jl

```diff
@@ -200,4 +198,23 @@ end
     @test findall(col -> eltype(col) <: Int, cols) == [1, 3]
 end
 
+@testset "data frame and grouped data frame iteration" begin
+    df = DataFrame(a=1:3, b=4:6, c=7:9)
+    dfv = @view df[1:3, 1:3]
+    gdf = groupby(df, :a)
+    gdfv = groupby(dfv, :a)
+
+    for x in (df, dfv)
+        @test collect(x) == map(identity, x) == [v for v in x] == [x[i, :] for i in 1:3]
```

Comment:
For now I did not implement custom `map` and `collect` and thus currently the following identity holds. Do we really think it is a problem?

note that e.g. `map` can take several iterables like `map(fun, df1, df2)` and the question is if we would want to intercept this also?

Reply:
I'd take a defensive approach and make everything for which we don't have a clear use case an error for now. People may expect `collect` to materialize a data frame in some way, like in Query.jl or dplyr, and be confused if they get a weird vector of rows. Likewise, we could imagine doing something useful in the future with a multi-argument `map`.

Regarding `map(identity, df)`, as I said at #2254, I think it should return a `DataFrame`, just like all cases where the function returns a row-like object.